### ISSUE TITLE
Error while running

### ISSUE THREAD

**Issue by yograjyog** [cutright/DVHA-MLCA#1]
Hi Dan.. First of all thanks for the code. While running the code I got this error? Your help is highly appreciated. 

![Capture](https://user-images.githubusercontent.com/17076501/85147864-0f886d80-b26d-11ea-8b0e-587fcd1d170e.JPG)

`

**Comment by cutright** [comment]
Looks like shapely isn’t installed?

Follow-up: did you install with pip or manually with setup.py? Did you get any errors when installing?

**Comment by SimonBiggs** [comment]
For this one, try:

pip install shapely>=1.7.0

Also, Dan, I would recommend pinning shapely to greater than or equal to
1.7.0 in your dvh package and this one. In 1.7.0 they released wheels that
make it installable on Windows without a world of pain. Putting that pin in
there will make it "just work" and will help solve these sorts of issues I
believe.

**Comment by cutright** [comment]
@yograjyog Any luck?

@SimonBiggs I've experienced some issues installing shapely >1.6.4.post2 on some environments. Although, I don't recall If 1.7.0 was out by then or not. But my intent with this software is to build PyInstaller binaries to avoid these issues.

**Comment by SimonBiggs** [comment]
Have you seen the embedded Python distributions?

https://www.python.org/ftp/python/3.8.3/python-3.8.3-embed-amd64.zip (from https://www.python.org/downloads/windows/).

They can be combined with something like NSIS (https://nsis.sourceforge.io/Main_Page) to create installers in cases where pyinstaller can be quite a pain...

**Comment by cutright** [comment]
I have not, but unfortunately, I've already learned those pains.  Seems to work pretty well with DVHA and DVHA DICOM Editor.  But if you're pointing to a true installer, that could be interesting.  I don't really like boot time with PyInstaller, particularly with Windows.

**Comment by yograjyog** [comment]
I have tried hard getting that work in windows. Then i shifted to Ubuntu.😀.. Now it is working.

**Comment by cutright** [comment]
Glad it's working, that's too bad about shapely on Windows.  [This](https://www.lfd.uci.edu/~gohlke/pythonlibs/#shapely) has been a good resource for shapely wheels too, if the ones on [PyPI](https://pypi.org/project/Shapely/) haven't been working.  I suspect if you were using the full  Anaconda, you wouldn't have issues... personally I stay away from anything Anaconda, too bloated for me.

I've had luck with something like: `pip install your_wheel_file.whl`

When I get time, I'll whip up a simple GUI and create a Windows and macOS binary to avoid this issue.

This code could pretty easily be adapted to other plan metrics, if you have any suggestions. I could  tweak the code a little bit to make it easy for outside contributors to add other metrics.

**Comment by yograjyog** [comment]
Thanks for your support. I will try your suggestion.  Is it possile to have Modulation Complexity Score adapted for VMAT. Your work will be highly appreciated. 

https://doi.org/10.1016/j.ejmp.2018.09.019

**Comment by cutright** [comment]
It should work for VMAT, at least it does on Pinnacle and Eclipse. Maybe you’re using Monaco and multi-arcs by chance? I think that’s the only beam type this code has issues with, which I’d be happy to accommodate with some anonymized examples.

**Comment by cutright** [comment]
Oh I see, I think this article has the formulas you are referring to?

Treatment plan complexity metrics for predicting IMRT pre-treatment quality assurance results
S.B.Crowe• T.Kairn• J.Kenny• R.T.Knight• B. Hill • C. M. Langton • J. V. Trapp
https://link.springer.com/content/pdf/10.1007/s13246-014-0274-9.pdf

**Comment by yograjyog** [comment]
Is that possible to have it within DVHA or MLCA?

**Comment by cutright** [comment]
For sure I'll add this to MLCA.

And worst case scenario, with respect to DVHA, it has a spreadsheet editor that lets you add new columns as well a copy and paste to/from Excel.  These new columns get stored in `.dvha` session files too.  That's how I generated my gamma pass-rate prediction models and TG-218 control charts.

Thanks for following up.

**Comment by cutright** [comment]
Looks like the work is already done? I came across this recently, but just noticed it’s the same author as the previous reference.
https://github.com/sbcrowe/pyradlib/tree/master/radlib/tada

**Comment by yograjyog** [comment]
Oh that sounds great. Thanks a lot for your effort.

**Comment by yograjyog** [comment]
I am novice.Trying very hard to learn python. How to use that program. A basic idea is highly appreciated. Thanks in advance.

**Comment by cutright** [comment]
I'll have to get back to you on that, it looks to me the library doesn't have a user interface? Even command line? I'm not able to really look at it right now though.

But from a python console, you can use the functions in the `complexity.py` and pass the pydicom object (e.g., plan, beam, etc.) into it.

So when his code asks for `plan` or `beam`, I'm assuming something like the following would get you there?
`plan = pydicom.read_file('rtplan.dcm')`
`beam = plan.BeamSequence[beam_index_of_interest]`

**Comment by yograjyog** [comment]
Thanks a lot.

**Comment by yograjyog** [comment]
I have tried something like this 

![complexity score input](https://user-images.githubusercontent.com/17076501/85879178-84bceb00-b7f7-11ea-994c-12838bb55f9e.JPG)


and got an output like this.

![complexity score](https://user-images.githubusercontent.com/17076501/85878856-0102fe80-b7f7-11ea-99ad-57f54481f264.JPG)

Instead of printing values it is showing the memory location and value.
Any suggestions? Your help is highly appreciated.

**Comment by cutright** [comment]
`_metric_list` is a dictionary of functions, but you'd like the values those functions return when passing in a pydicom object.

I think the intent is to do something like:
```
import pydicom
from complexity import plan_complexity
plan = pydicom.read_file('rtplan.dcm')
metrics_to_calc = ['MCS', 'MAD']  # fill this list with the calculations you want
results = plan_complexity(plan, metrics_to_calc )
```

Unfortunately that doesn't work. He's looking for a tag name of `BeamLimitingDevices` which I'm unfamiliar with and doesn't exist in my plan files.  There is also some issues with `pandas`, perhaps the latest version breaks his code... or there are some other DICOM tag issues.  Your best bet is to contact him in the meantime.

**Comment by cutright** [comment]
Got it!  I think maybe pydicom had some changes since he wrote this? There's a good chance you could just downgrade your `pydicom`, but rather than guessing and checking, I changed attributes like `BeamLimitingDevices` to `BeamLimitingDeviceSequence`... similiarly for `ControlPoints`, `BeamLimitingDevicePositions`, etc.

Use [this](https://github.com/cutright/pyradlib/blob/master/radlib/tada/complexity.py) for complexity.py.  Then do the following code.  
(**Edited by linking the code to a file, rather than posting in the comment**)

```
import pydicom
from complexity import plan_complexity
ds = pydicom.read_file('rtplan.dcm')
results = plan_complexity(ds, ['MCS'])
print(results)
      Name   MU      Dose       MCS
1   01 CCW  236   0.58091   0.32281
2    02 CW  263   0.71527  0.267371
3  03 CCW2  208  0.528058  0.385578
4   04 CCW  431  0.954637  0.272984
5    05 CW  407  0.830187  0.229794
```

**Comment by SimonBiggs** [comment]
Probably worth making this a pull request...

**Comment by yograjyog** [comment]
Thanks @cutright for spending your valuable time and helping me. To what version of pydicom i want to downgrade? Is there any particular version number?

**Comment by yograjyog** [comment]
Thanks a lot spending your time and helping me. I can able to calculate MCS alone and remaining metrics are throwing errors. As you suggested i will downgrade the pydicom and give a try, but is there any particular version i need to downgrade? Thanks again for your support.

**Comment by cutright** [comment]
There are a number of issues with this code, I'm afraid doing a pull request will open a giant can of worms.  It also looks to me like the code was only partially updated for `pydicom 1.0`.  He has `import pydicom` statements (instead of `import dicom`) which is a `1.0` change, but I have to downgrade to `0.9.9` for the previous tag names I mentioned to be available.

My suggestion is to contact the developer directly and see what environment he used.  It looks like he has relatively good coding practices, my guess is that it works very nicely if you have the right environment.  Although, you may find some TPS-specific issues still.  For example, my Pinnacle step-N-shoot plans crash with MCS:  
`RuntimeWarning: divide by zero encountered in double_scalars
  return (lsv_left / (exposed_open_leaf_pairs * min_max_left)) * (lsv_right / (exposed_open_leaf_pairs * min_max_right))`

I wish I had the bandwidth to dive in deeper right now.

I also posted my changes to his code on a branch [here](https://github.com/cutright/pyradlib/blob/master/radlib/tada/complexity.py).

**Comment by cutright** [comment]
@sbcrowe Hi Scott, maybe you can help us out here?  Would you mind posting your `pip list` for pyradlib (assuming you used pip)?

**Comment by yograjyog** [comment]
Thanks for the suggestion. In MLCA also i was having a problem like, 

While i am running the MLCA <init scanning dir>, it is showing that it is analyzing but it is not displaying the results. (My environment is Ubuntu)

**Comment by cutright** [comment]
Can you be more specific with some code or terminal output?

**Comment by yograjyog** [comment]
I tried with all the codes that you have mentioned in  MLCA  readme

**Comment by yograjyog** [comment]
> I tried with all the codes that you have mentioned in MLCA readme. 

Right now I dont have access to the system. May be tomorrow i will send the screenshot

**Comment by sbcrowe** [comment]
> @sbcrowe Hi Scott, maybe you can help us out here? Would you mind posting your `pip list` for pyradlib (assuming you used pip)?

That code was initially written for an earlier version of pydicom, and I no longer have the environment. Even if properly updated for `pydicom 1.0`, I'm confident the code will fail inelegantly with variations in DICOM attributes from different planning systems, different MLC systems, etc. that I wasn't expecting. That is, I don't think it is ready for use by someone not in a position to modify it to work for their use case.

**Comment by SimonBiggs** [comment]
@sbcrowe would there be any interest in potentially merging this work into either DVHA-MLCA or pymedphys-labs (https://github.com/pymedphys/pymedphys/tree/master/pymedphys/labs)?

**Comment by yograjyog** [comment]
> Can you be more specific with some code or terminal output?

![Screenshot from 2020-06-30 11-26-40](https://user-images.githubusercontent.com/17076501/86088849-b6b6a180-bac4-11ea-80fc-f06e16dd4ef7.png)

**Comment by cutright** [comment]
And there’s no CSV file created?

**Comment by yograjyog** [comment]
> And there’s no CSV file created?

No it is not creating the csv.

**Comment by yograjyog** [comment]
> @sbcrowe would there be any interest in potentially merging this work into either DVHA-MLCA or pymedphys-labs (https://github.com/pymedphys/pymedphys/tree/master/pymedphys/labs)?

@SimonBiggs Is there any way to try this out?

**Comment by SimonBiggs** [comment]
Hi @yograjyog,

@db72-git recently committed some plan complexity tools to the labs of pymedphys:

https://github.com/pymedphys/pymedphys/tree/master/pymedphys/labs/plancomplexity 

He would be the best person to discuss with regarding tools that are already in the pymedphys labs.

Regarding having @sbcrowe's code merged into the labs, I do require that any code merged into the labs has someone designate themselves as responsible for that code. A "Code Owner" so to speak. Even though @sbcrowe's code is open source, I won't be merging it into pymedphys without his permission as I only want code merged into pymedphys that someone is willing to take a degree of responsibility for.

So, in short, without some more work undergone, no, there is not currently an easy way for you to try it out.

Cheers,
Simon

**Comment by db72-git** [comment]
Hi all,

Sorry, I have been super busy lately so not got round to looking at this any further yet. I did indeed upload some complexity code to the pymedphys labs. The next thing to do with this is to make it run from the pymedphys command line which I have not done yet. So there are a few more stages I reckon before it's anything that you could try out easily but we're getting there. 

It's also pretty experimental at the moment and I can only guarantee it works with Pinnacle plans but will be happy to modify it to work for others if it needs it.

**Comment by cutright** [comment]
@yograjyog The fastest way to resolve this is if you could provide some anonymized files that exhibit this issue.  It's very peculiar that no CSV is created, it should at least create a CSV with only the column headers.  But perhaps there's some OS permission issues?

There is only one try/except clause (shown below), so the code is not crashing somewhere and being ignored.
https://github.com/cutright/DVHA-MLCA/blob/5d45edf8dd21781d8dfe721342ac7fa8721b0e22/mlca/main.py#L47-L52

In the mean time, I'm working on a way to more easily call the code from a python console as opposed to calling from a terminal... and adding a verbose mode so the results can be printed directly to the terminal.

**Comment by yograjyog** [comment]
> @yograjyog The fastest way to resolve this is if you could provide some anonymized files that exhibit this issue. It's very peculiar that no CSV is created, it should at least create a CSV with only the column headers. But perhaps there's some OS permission issues?
> 
> There is only one try/except clause (shown below), so the code is not crashing somewhere and being ignored.
> 
> https://github.com/cutright/DVHA-MLCA/blob/5d45edf8dd21781d8dfe721342ac7fa8721b0e22/mlca/main.py#L47-L52
> 
> In the mean time, I'm working on a way to more easily call the code from a python console as opposed to calling from a terminal... and adding a verbose mode so the results can be printed directly to the terminal.

I have uninstalled and reinstalled, now it is creating the .csv file Thanks for your support.

**Comment by yograjyog** [comment]
Hi Dan. Please have a look in to this script . Appreciate if you could add this to MLCA..  

[https://github.com/Admodan/Research](url)

**Comment by yograjyog** [comment]
Hi Dan! 

Is it possible for you to help me running the following code. I asked the code author but couldnt get the success. I appreciate if you can help me. Thanks in advance.

https://github.com/daolamgit/Predaq.git
